### PR TITLE
[Feature-9467] add DAMENG DataSource

### PR DESCRIPTION
**Pull request by lenian**
## Verify this pull request
https://github.com/apache/dolphinscheduler/issues/9467

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-dm/src/main/java/org/apache/dolphinscheduler/plugin/datasource/dm/param/DmConnectionParam.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import org.apache.dolphinscheduler.spi.datasource.BaseConnectionParam;
+
+public class DmConnectionParam extends BaseConnectionParam {
+
+    @Override
+    public String toString() {
+        return "DmConnectionParam{"
+                + "user='" + user + '\''
+                + ", password='" + password + '\''
+                + ", address='" + address + '\''
```
Comment (DarkAssassinator) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-dm/src/main/java/org/apache/dolphinscheduler/plugin/datasource/dm/param/DmConnectionParam.java]: IMHO, remove `password` in `toString()` should be better, this prevents the password from being printed to the log file. WDYT.

Comment (lenian) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-dm/src/main/java/org/apache/dolphinscheduler/plugin/datasource/dm/param/DmConnectionParam.java:29]: ok, I'll remove password in toString()

Comment (lenian) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-dm/src/main/java/org/apache/dolphinscheduler/plugin/datasource/dm/param/DmConnectionParam.java]: The `toString()` method of the `OracleConnectionParam`、`MySQLConnectionParam `class has the password attribute printed.I modified it according to the mysql and oracle specifications。Now I need to remove `password `in `toString()`?

Comment (kezhenxu94) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-dm/src/main/java/org/apache/dolphinscheduler/plugin/datasource/dm/param/DmConnectionParam.java]: > The `toString()` method of the `OracleConnectionParam`、`MySQLConnectionParam `class has the password attribute printed.I modified it according to the mysql and oracle specifications。Now I need to remove `password `in `toString()`?

Please remove all `password` in those classes, also, consider use `lombok` `ToString(excludes = ` to simplify the codes

File: dolphinscheduler-bom/pom.xml [hunk 2]
```diff
@@ -380,6 +381,13 @@
                 <scope>test</scope>
             </dependency>
 
+            <dependency>
+                <groupId>com.dameng</groupId>
+                <artifactId>DmJdbcDriver18</artifactId>
+                <version>${dm-jdbc.version}</version>
+                <scope>test</scope>
```
Comment (kezhenxu94) [inline dolphinscheduler-bom/pom.xml]: Are you sure we want to use `test` scope for this dependency? Users who want to use dm plugin have to download the jar and put it in the correct folder. I think `DmJdbcDriver18` is Apache license so we can use it safely in our distribution package

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-dm/src/main/java/org/apache/dolphinscheduler/plugin/datasource/dm/param/DmConnectionParam.java (hunk 1 above)
Comment (lenian) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-dm/src/main/java/org/apache/dolphinscheduler/plugin/datasource/dm/param/DmConnectionParam.java]: ok, I'll remove `password` in `toString()`

File: dolphinscheduler-bom/pom.xml (hunk 2 above)
Comment (lenian) [reply dolphinscheduler-bom/pom.xml]: I checked  DmJdbcDriver18 is Apache License  2.0, I want the user to use directly, without adding the driver jar package. I'll remove the test scope.

**Review by SbloodyS — COMMENTED** [review]
Please add some english docs.

**Comment by lenian** [comment]
> Please add some english docs.

Ok, I will submit the English document

**Comment by lenian** [comment]
Please add lable for pr, thanks.

**Comment by lenian** [comment]
> Please add some english docs.

Please add lable for pr, thanks.

**Comment by caishunfeng** [comment]
Hi @lenian please resolve the conflicts.

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/DruidDataSourceClient.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        // Checking data source client
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        try {
+            this.jdbcTemplate.execute(this.baseConnectionParam.getValidationQuery());
+        } catch (Exception e) {
+            throw new RuntimeException("JDBC connect failed", e);
+        } finally {
+            logger.info("Time to execute check jdbc client with sql {} for {} ms ",
+                    this.baseConnectionParam.getValidationQuery(), stopwatch.elapsed(TimeUnit.MILLISECONDS));
+        }
```
Comment (caishunfeng) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/DruidDataSourceClient.java]:
```suggestion
        try {
            this.jdbcTemplate.execute(this.baseConnectionParam.getValidationQuery());
            logger.info("Time to execute check jdbc client with sql {} for {} ms ",
                    this.baseConnectionParam.getValidationQuery(), stopwatch.elapsed(TimeUnit.MILLISECONDS));
        } catch (Exception e) {
            throw new RuntimeException("JDBC connect failed", e);
        } 
```

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/DruidDataSourceClient.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    @Override
+    public Connection getConnection() {
+        try {
+            return this.dataSource.getConnection();
+        } catch (SQLException e) {
+            logger.error("get druidDataSource Connection fail SQLException: {}", e.getMessage(), e);
```
Comment (caishunfeng) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/DruidDataSourceClient.java]:
```suggestion
            logger.error("get druidDataSource Connection fail SQLException: ", e);
```

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-dm/src/main/java/org/apache/dolphinscheduler/plugin/datasource/dm/param/DmDataSourceProcessor.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    @Override
+    public String getValidationQuery() {
+        return DataSourceConstants.DM_VALIDATION_QUERY;
+    }
+
+    @Override
+    public String getJdbcUrl(ConnectionParam connectionParam) {
+        DmConnectionParam dmConnectionParam = (DmConnectionParam) connectionParam;
+        String jdbcUrl = dmConnectionParam.getJdbcUrl();
+        if (!StringUtils.isEmpty(dmConnectionParam.getOther())) {
```
Comment (caishunfeng) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-dm/src/main/java/org/apache/dolphinscheduler/plugin/datasource/dm/param/DmDataSourceProcessor.java]:
```suggestion
        if (StringUtils.isNotEmpty(dmConnectionParam.getOther())) {
```

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/DruidDataSourceClient.java (hunk 3 above)
Comment (lenian) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/DruidDataSourceClient.java]: ok, I will modify it according to your suggestion

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/DruidDataSourceClient.java (hunk 4 above)
Comment (lenian) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/DruidDataSourceClient.java]: ok, I will modify it according to your suggestion

**Comment by davidzollo** [comment]
image
please solve these conflicts

**Comment by lenian** [comment]
> 

I have resolved the code conflict

**Comment by davidzollo** [comment]
good job
Please use dameng instead of dm, nobody knows what `dm` means
Please add license file refer to this doc, https://dolphinscheduler.apache.org/en-us/docs/latest/user_doc/contribute/join/DS-License.html

**Comment by lenian** [comment]
> good job Please use dameng instead of dm, nobody knows what `dm` means Please add license file refer to this doc, https://dolphinscheduler.apache.org/en-us/docs/latest/user_doc/contribute/join/DS-License.html
I will improve it according to your suggestion

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/pom.xml [hunk 6]
```diff
@@ -141,5 +141,10 @@
             <groupId>com.zaxxer</groupId>
             <artifactId>HikariCP</artifactId>
         </dependency>
+
+        <dependency>
+            <groupId>com.alibaba</groupId>
+            <artifactId>druid</artifactId>
+        </dependency>
     </dependencies>
```
Comment (CalvinKirs) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/pom.xml]: Why are you introducing a new database connection pool? we already have `HikariCP`.

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-dameng/src/main/java/org/apache/dolphinscheduler/plugin/datasource/dameng/param/DamengDataSourceProcessor.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        DamengConnectionParam damengConnectionParam = new DamengConnectionParam();
+        damengConnectionParam.setJdbcUrl(jdbcUrl);
+        damengConnectionParam.setDatabase(damengConnectionParam.getDatabase());
+        damengConnectionParam.setAddress(address);
+        damengConnectionParam.setUser(dmDatasourceParam.getUserName());
+        damengConnectionParam.setPassword(PasswordUtils.encodePassword(dmDatasourceParam.getPassword()));
+        damengConnectionParam.setDriverClassName(getDatasourceDriver());
+        damengConnectionParam.setValidationQuery(getValidationQuery());
+        damengConnectionParam.setOther(transformOther(dmDatasourceParam.getOther()));
+        damengConnectionParam.setProps(dmDatasourceParam.getOther());
```
Comment (ruanwenjun) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-dameng/src/main/java/org/apache/dolphinscheduler/plugin/datasource/dameng/param/DamengDataSourceProcessor.java]: Please don't use `setProps`, I remember we don't have props in `BaseDataSourceParamDTO`, you need to rebase the upstream code.

Comment (lenian) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-dameng/src/main/java/org/apache/dolphinscheduler/plugin/datasource/dameng/param/DamengDataSourceProcessor.java]: Ok, I have rebase BaseDataSourceParamDTO

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/pom.xml (hunk 6 above)
Comment (lenian) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/pom.xml]: HikariCP does not support Dameng, druid supports Dameng

Comment (CalvinKirs) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/pom.xml]: I don't understand, it's just a data source connector management, why doesn't it support dm, can you explain why?

Comment (lenian) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/pom.xml]: thanks，I just tested the DaMeng database with HikariCP，HikariCP supports DaMeng

**Review by Tianqi-Dotes — DISMISSED** [review]
LGTM, thanx for the efforts.

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/pom.xml [hunk 8]
```diff
@@ -141,5 +141,6 @@
             <groupId>com.zaxxer</groupId>
             <artifactId>HikariCP</artifactId>
         </dependency>
+
```
Comment (SbloodyS) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/pom.xml]: Please remove unnessnary change.

Comment (lenian) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/pom.xml]: Ok, I'll remove the unnecessary changes right away

**Review by Tianqi-Dotes — DISMISSED** [review]

File: docs/docs/zh/guide/datasource/dameng.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# DAMENG数据源
+
+![dameng](../../../../img/new_ui/dev/datasource/dameng_cn.png)
```
Comment (SbloodyS) [inline docs/docs/zh/guide/datasource/dameng.md]: Please use English screenshot, we will not maintain screenshots in different languages.

Comment (lenian) [reply docs/docs/zh/guide/datasource/dameng.md]: Ok, I will take a screenshot in English

**Review by SbloodyS — APPROVED** [review]
+1

**Review by Tianqi-Dotes — APPROVED** [review]

**Pull request merged**